### ISSUE TITLE
Sound effects are noises in PAL(Chinese RPG) with Intel HDA

### ISSUE THREAD

**Issue by Jimmy-Z** [crazii/SBEMU#148]
First of all, thank you for this amazing software!

The game is probably not very well known in the west but it's the most prominent DOS game in China back in the late 90s.
https://en.wikipedia.org/wiki/The_Legend_of_Sword_and_Fairy_(video_game)

Adlib music works fine, but various sound effects sound like white noise (with a little variation, if that makes sense), like a while after you start a new game, after some dialogue your aunt will hit you with a pan, and all battle sound effects (needs to proceed a bit further into the game).

I tried this on two of my older laptops, both with Intel HDA, both had this issue.

Then I tried this in 86box with Sound Blaster PCI 128 (ES1373) with sbemu, sound effects works fine.

1.0.0 beta 5 on a bare minimum of freedos 1.3, settings are according to readme.txt.

**Comment by Jimmy-Z** [comment]
update: I did more tinkering

* laptop is thinkpad x230, cpu i5-3210m, intel hda.
* freedos 1.3 on a usb stick.
* other popular games like doom, quake, duke nukem 3d seems to be working fine.
    * I only tested shareware/demo versions, so quake doesn't have cd music, but sound effects sounds fine.
* /K44100, no effect
* /FIXTC, no effect
* /T6, no effect
* /I5, no effect
    * as a side note, in hwinfo:
        * 5 is listed as free(LPT2)
        * 7 is used (onboard ethernet, not plugged, tried to disable it in bios/uefi setting, doesn't work)
        * 9 is used (rerouted irq 2)
* reload bios default, no effect
* /I9, worse, music still works, but sound effects is _disabled_, not even noise.
    * other games I tried to test doesn't support setting irq to 9.
* use himemx.sys instead of jemmex, and disable real mode support, even worse, music is gone.
    * ~I don't quite understand, this is supposedly a protected mode game~, but I could be wrong on this.
        * update: it's likely a real mode game.
    * doom, quake, duke nukem 3d still works.
* descent setup.exe, this is the only other thing I tested that sound effects also doesn't work.
    * in setup.exe, select "Test Digital Sound Card", I can hear noise coming out on the center/left/right channel respectively
        * on dosbox it's a woman's voice saying "seven"
    * in game sound effects sounds ok, however
    * I'm sure when I tested this for the first few times, it worked, but after some tinkering it now doesn't, I'm pretty sure I've reverted all settings, I don't know, maybe I should doubt my memory.
* descent 2, with a setup.exe looking very similar to 1, however, it works fine, I can hear the woman's voice saying "zero" clearly.
* /VOL9, since default volume is so low, but I can hear the "seven" in descent setup.exe vaguely covered under the noise now.
* after all these tinkering, the noise _pattern_ is different now, and I can't seem to get it back to the "white noise" pattern.

**Comment by volkertb** [comment]
Thanks for the extensive testing, @Jimmy-Z! 🙏🏽

I know this project could do with more active development lately. I'm trying to contribute some, but my expertise doesn't come even close to that of @crazii. I hope to find more time to get deeper involved in this project in my free time, but I also hope that other developers step up to help maintain and improve SBEMU.

In the meantime, could you maybe test this game with older builds of SBEMU as well, to see if there are any regressions in this release? Thanks again!

**Comment by Jimmy-Z** [comment]
Good idea! I'll test this, maybe tomorrow.

**Comment by Jimmy-Z** [comment]
I tested 1.0-beta4 and UserBuild_2024.04.29_09-29, unfortunately both exactly the same symptom, what other versions do you think worth testing?

**Comment by volkertb** [comment]
I know it's quite an old version, but could you maybe test this with [1.0beta3](https://github.com/crazii/SBEMU/releases/tag/1.0beta3)? Earlier this month, someone posted [a message](https://www.vogons.org/viewtopic.php?p=1312254#p1312254) in the SBEMU thread on Vogons, mentioning that YMF724 cards stopped working after beta3 as well. Not sure if that's related to your issues with Intel HDA, but let's see if these issues are correlated.

There are a lot of snapshot releases between beta 3 and beta 4, but ff you can confirm that it's still working with that version, then we can at least start bisecting to get to the version (and with that hopefully also the change) that introduced this bug .

No rush, though. We're in the midsts of the winter holidays, after all. Whenever you get around to it and feel like diving into it is fine. Thanks again, enjoy the holidays and have a wonderful and healthy 2025! :relaxed:

**Comment by Jimmy-Z** [comment]
Thanks for the suggestion, and happy holidays to you too! ☺️

**Comment by Jimmy-Z** [comment]
Hi, it's been half a year and I remembered this, but 1.0beta3 behaves exactly the same.

**Comment by Jimmy-Z** [comment]
Baron-von-Riedesel figured out this is a game bug and provided a workaround: https://github.com/Baron-von-Riedesel/VSBHDA/issues/39#issuecomment-3067195969

After applying the workaround sound effects (digital sounds) are no longer noise, but there are some popping/cracking, which didn't appear in VSBHDA.

**Comment by crazii** [comment]
this bug is fixed with VDPMI: https://www.vogons.org/viewtopic.php?p=1418209#p1418209
PAL uses XMS for sound FX, the problem is XMM (himem.sys or equivalent) reported some XMS and PAL assumes it to be at lower 16M, which is most common case but not always true, because HDPMI and other program already uses the lower 16M XMS.
VDPMI on the other hand, reserved the lower 16M during loading, make the space for legacy games.

It is just weird to report 16M memory, but the real allocation is above 16M anyways.

**Comment by crazii** [comment]
the bug is fixed in an very early beta of VDPMI here:  https://www.vogons.org/viewtopic.php?p=1258563#p1258563 , which I forgot to mention it.
I am very sorry for the delayed reply that I didn't read the issues. I was busy and gone offline for a long time.

BTW, I played PAL in the 90s when I was in elementary school. Yes, I am in China too.

**Comment by palxex** [comment]
By the way, has anyone tested the new `vmpu` feature with the original DOS version of PAL?

I was adding MPU-401 support to the DOS version of SDLPal and used the original game as a reference when comparing MPU-401 behavior. During this process, I tried PAL with SBEMU's `vmpu` (launched with `/vmpu` and placing `sbemusf.sf2` to same folder of sbemu.exe, which was simply converted from `gm.dls` used by the Microsoft GS Wavetable Synth), but the game always hangs with a black screen right after the Softstar trademark animation. According to my tracing, no MIDI data is sent before this point, which suggests the hang happens when the game starts sending MIDI events.

I tested both the HDPMI and VDPMI versions and got the same result.

For comparison, the same `PAL.exe` plays MIDI correctly with VSBHDA using the `/sf:` option with the same SF2.

It would be helpful to know whether this is a known compatibility issue or something specific to my setup.

**Comment by crazii** [comment]
> By the way, has anyone tested the new `vmpu` feature with the original DOS version of PAL?
> 
> I was adding MPU-401 support to the DOS version of SDLPal and used the original game as a reference when comparing MPU-401 behavior. During this process, I tried PAL with SBEMU's `vmpu` (launched with `/vmpu` and placing `sbemusf.sf2` to same folder of sbemu.exe, which was simply converted from `gm.dls` used by the Microsoft GS Wavetable Synth), but the game always hangs with a black screen right after the Softstar trademark animation. According to my tracing, no MIDI data is sent before this point, which suggests the hang happens when the game starts sending MIDI events.
> 
> I tested both the HDPMI and VDPMI versions and got the same result.
> 
> For comparison, the same `PAL.exe` plays MIDI correctly with VSBHDA using the `/sf:` option with the same SF2.
> 
> It would be helpful to know whether this is a known compatibility issue or something specific to my setup.

Thank you for the feedback! It will be tested.

**Comment by crazii** [comment]
I tested the original PAL and SBEMU stuck on dead loops in VMPU_Process_Messages, where *temp_buffer is 0x3A, @Cacodemon345 any ideas?
BTW Previously I only tested duke3d and doom.

**Comment by Cacodemon345** [comment]
This sounds like it's either failing to insert the running status into the buffer for whatever reason, or failing to properly process SysEx messages, causing invalid messages to be processed. Workaround would be to force running status insertion into the message buffer, and silently ignoring invalid messages if at all possible.

**Comment by crazii** [comment]
> This sounds like it's either failing to insert the running status into the buffer for whatever reason, or failing to properly process SysEx messages, causing invalid messages to be processed. Workaround would be to force running status insertion into the message buffer, and silently ignoring invalid messages if at all possible.

Sounds like that. I'm now debugging it and here is a typical sequence of PAL:

```
vmpu.c    (0174):VMPU_Write(330)=F7
vmpu.c    (0174):VMPU_Write(330)=C1
vmpu.c    (0174):VMPU_Write(330)=48
vmpu.c    (0174):VMPU_Write(330)=B1
vmpu.c    (0174):VMPU_Write(330)=7
[f0 a 41 10 42 12 40 0 7f 0 41 f7 c1 48 ]vmpu.c    (0065):MIDI:f0, i=0, a=14
vmpu.c    (0065):MIDI:c1, i=12, a=14
vmpu.c    (0174):VMPU_Write(330)=64
vmpu.c    (0174):VMPU_Write(330)=B1
vmpu.c    (0174):VMPU_Write(330)=A
vmpu.c    (0174):VMPU_Write(330)=14
vmpu.c    (0174):VMPU_Write(330)=E1
vmpu.c    (0174):VMPU_Write(330)=0
[b1 7 64 b1 a 14 ]vmpu.c    (0065):MIDI:b1, i=0, a=6
vmpu.c    (0065):MIDI:b1, i=3, a=6
vmpu.c    (0174):VMPU_Write(330)=91
vmpu.c    (0174):VMPU_Write(330)=4A
vmpu.c    (0174):VMPU_Write(330)=F
[e1 0 91 ]vmpu.c    (0065):MIDI:e1, i=0, a=3
vmpu.c    (0174):VMPU_Write(330)=B1
vmpu.c    (0174):VMPU_Write(330)=40
vmpu.c    (0174):VMPU_Write(330)=3A
```

Here the last sequence is (E1, 0, 91, 4A, F), (B1,40,3A)
IIUC, the status E1 should be inserted so that it be (E1, 0, 91, E1, 4A, F)? or the 4A,F being ignored?

**Comment by Cacodemon345** [comment]
The former to avoid message loss issues.

**Comment by crazii** [comment]
> The former to avoid message loss issues.

Thanks, but is 91 supposed to be a status byte? so theoretically, there seems a data missing after E1?
it should be E1, 0, X, 91, 4A, F

**Comment by palxex** [comment]
Just tried modifying DOSBox-X and captured the MPU-401 data that PAL.EXE sends. It seems it always sends pitch bend events (0xE0 | channel) with only one data byte, so seems the MSB is lost (as can be seen in the MIDI file). 
Another problem is that it wrongly sends SysEx without stripping the SMF length byte, but that does not block the loop.

[midi_data.zip](https://github.com/user-attachments/files/31330197/midi_data.zip)

[dosbox-capture-mpu401.patch](https://github.com/user-attachments/files/31330217/dosbox-capture-mpu401.patch)

**Comment by Cacodemon345** [comment]
I think that's very puzzling for me too. I think it's better treated as a status byte.

**Comment by crazii** [comment]
> Just tried modifying DOSBox-X and captured the MPU-401 data that PAL.EXE sends. It seems it always sends pitch bend events (0xE0 | channel) with only one data byte, so seems the MSB is lost (as can be seen in the MIDI file). Another problem is that it wrongly sends SysEx without stripping the SMF length byte, but that does not block the loop.
> 
> [midi_data.zip](https://github.com/user-attachments/files/31330197/midi_data.zip)
> 
> [dosbox-capture-mpu401.patch](https://github.com/user-attachments/files/31330217/dosbox-capture-mpu401.patch)

thank, I added data insertion, can you test this one:

[SBEMU_VDPMI.zip](https://github.com/user-attachments/files/31343141/SBEMU_VDPMI.zip)

As I tested, the SoundFX is broken in MPU mode, and using the Yamaha XG sound font generates weird tones.
UPDATE: broken SFX fixed.

**Comment by palxex** [comment]
> thank, I added data insertion, can you test this one:
> 
> [SBEMU_VDPMI.zip](https://github.com/user-attachments/files/31332043/SBEMU_VDPMI.zip)
> 
> As I tested, the SoundFX is broken in MPU mode

Thanks for the build! In my test, the VOC sound effects are also broken, so I guess this is a separate issue.
I can confirm that PAL can now play MIDI without hanging with this version, although it sometimes sounds a bit strange. I don't have a very good ear for MIDI myself, but one of my friends does. He's having trouble running the VDPMI version on his machine. Could you please compile an HDPMI version too? Or just the modified source would be fine.

> and using the Yamaha XG sound font generates weird tones.

Since PAL's MIDI was composed for the Roland SC-55 (according to this [video interview](https://www.bilibili.com/video/BV1QnBjB4E7B) with its music composer), maybe it would be better to test it with a GS sound font. 
[SBEMUSF.zip](https://github.com/user-attachments/files/31333246/SBEMUSF.zip)

**Comment by crazii** [comment]
> > thank, I added data insertion, can you test this one:
> > [SBEMU_VDPMI.zip](https://github.com/user-attachments/files/31332043/SBEMU_VDPMI.zip)
> > As I tested, the SoundFX is broken in MPU mode
> 
> Thanks for the build! In my test, the VOC sound effects are also broken, so I guess this is a separate issue. I can confirm that PAL can now play MIDI without hanging with this version, although it sometimes sounds a bit strange. I don't have a very good ear for MIDI myself, but one of my friends does. He's having trouble running the VDPMI version on his machine. Could you please compile an HDPMI version too? Or just the modified source would be fine.
> 
> > and using the Yamaha XG sound font generates weird tones.
> 
> Since PAL's MIDI was composed for the Roland SC-55 (according to this [video interview](https://www.bilibili.com/video/BV1QnBjB4E7B) with its music composer), maybe it would be better to test it with a GS sound font. [SBEMUSF.zip](https://github.com/user-attachments/files/31333246/SBEMUSF.zip)

the SC-55 sf seems better but still weird, I'll make a HDPMI build tomorrow.
I also tested dosbox-x with no extra midi configs, it seems worse.

**Comment by Cacodemon345** [comment]
TinySoundFont does not actually support modulators in the sound fonts for the record because it'd slow it down on slow ARM devices, so that might cause some soundfonts to sound strange.

**Comment by palxex** [comment]
> the SC-55 sf seems better but still weird, I'll make a HDPMI build tomorrow. I also tested dosbox-x with no extra midi configs, it seems worse.

Thanks! I'm just curious what you mean by “worse” with DOSBox-X. The MIDI quality of PAL is generally considered not as good as its FM music anyway, so for me the main goal here was just to get it working without hanging :)

BTW, there are two things that might be useful for comparison:

1. If the strange sound is related to the sustain pedal not being handled well, there is actually a [fan-made modified version](https://github.com/user-attachments/files/31335379/midi-diudiumod.zip) of `MIDI.MKF` that fixes this issue. It also doesn't have the infinite-loop problem, even with the previous version of SBEMU, since it eliminates all pitch-bend events as well. But anyway, it's just a fan-made modification, so we can only use it as a reference.

2. As far as I know, incomplete pitch-bend messages are generally just discarded rather than completed with a missing byte. DOSBox does the same, and this behavior has apparently been tolerated by hardware and software implementations for a long time. Maybe it would be worth a try as well, just to see if it sounds closer to what people are used to hearing.

**Comment by crazii** [comment]
> TinySoundFont does not actually support modulators in the sound fonts for the record because it'd slow it down on slow ARM devices, so that might cause some soundfonts to sound strange.

Thanks,
Is it possible to add it?  Or is there any forks having done this?

**Comment by crazii** [comment]
> > the SC-55 sf seems better but still weird, I'll make a HDPMI build tomorrow. I also tested dosbox-x with no extra midi configs, it seems worse.
> 
> Thanks! I'm just curious what you mean by “worse” with DOSBox-X. The MIDI quality of PAL is generally considered not as good as its FM music anyway, so for me the main goal here was just to get it working without hanging :)
> 
> BTW, there are two things that might be useful for comparison:
> 
> 1. If the strange sound is related to the sustain pedal not being handled well, there is actually a [fan-made modified version](https://github.com/user-attachments/files/31335379/midi-diudiumod.zip) of `MIDI.MKF` that fixes this issue. It also doesn't have the infinite-loop problem, even with the previous version of SBEMU, since it eliminates all pitch-bend events as well. But anyway, it's just a fan-made modification, so we can only use it as a reference.
> 2. As far as I know, incomplete pitch-bend messages are generally just discarded rather than completed with a missing byte. DOSBox does the same, and this behavior has apparently been tolerated by hardware and software implementations for a long time. Maybe it would be worth a try as well, just to see if it sounds closer to what people are used to hearing.

Thanks, Google AI told me MIDI softwares discard all incomplete messages. I will give it a try.

I remember replacing the original mkf with fan modified one, multiple times for different copies during the years, but I'm not sure this copy I tested is the original or modified.
In DOSBox-X there is a low pitch sound in the background keep playing when starting a new game.

**Comment by crazii** [comment]
About the broken SFX, it's again the problem of the topic of this issue.
it seems HimemX allocates memory in a ping-poing pattern, it is an expected behavior of HimemX as its readme states "'wild' mix" while HimemX2 allocates memory "from low addresses to high addresses"
Here is the allocation log of PAL, left: SB without MIDI(MPU) right: SB with MPU, the last allocation address is to high for PAL (0x52b3c00).

<img width="1233" height="408" alt="Image" src="https://github.com/user-attachments/assets/521ca745-8d5b-4b34-97ac-8c35c37d6554" />

I tried HimemX2 sometime ago and again just now.  but HimemX2 doesn't use the low memory blocks (in MBs) freed by VDPMI, but alloc from the end of VDPMI's memory region.

Currently for VDPMI the only fix is to take over & implement the xms2 allocation functions but I don't think it's the right way (may be good for VDS in the future).
Maybe changes the behavior of HimemX2 should be better.

**Comment by crazii** [comment]
Here's the HDPMI edition of SBEMU, as I tested it does not freeze on MPU:
also incomplete MIDI messages are now discarded.

this package contains a single binary only.
[sbemu.zip](https://github.com/user-attachments/files/31343108/sbemu.zip)

**Comment by crazii** [comment]
I checked the HIMEM.SYS and it allocates memory with high (above16M) & low (below 16M) address in a ping-pong style like HimemX,
now a dirty hack is added to retry the allocation to make sure it's below 16M. and the broken SFX  is fixed. binary updated in here: 
MPU message processing also discards incomplete message in this build, but I failed to see any difference of the "0 insertion" build.
https://github.com/crazii/SBEMU/issues/148#issuecomment-5379902356

This issue (of this topic, i.e. SFX problem) is finally totally fixed (although in some dirty way), and I will close it later.

**Comment by palxex** [comment]
Thanks for new builds! Below is my results, will update my friend's when available
On VirtualBox: HDPMI ver works well on MIDI and SFX. VDPMI ver works well on MIDI, but SFX cannot be verified since it will crash instantly after game open/start  a new save slot( the softstar logo/splash screen/select screen ok).

<img width="3839" height="2055" alt="Image" src="https://github.com/user-attachments/assets/12291cab-f7df-4209-b671-51d3dcfb8eee" />

On My PC, FreeDOS 1.3( but fdnpkg upgraded to lastest packages ) with CSM: on both DPMI MIDI works well , but SFX broken ( verifed same binary with virtualbox ) . Both tested on HDA and CMI8738, same result.
Additional info: if I run `xmsres /l 15` before pal.exe, the result will change, more like vbox situation: SFX will work again with HDPMI ver, and game will crash with VDPMI ver, and same Error 05 will be reported( only the seg addr was changed ).

**Comment by crazii** [comment]
> Thanks for new builds! Below is my results, will update my friend's when available On VirtualBox: HDPMI ver works well on MIDI and SFX. VDPMI ver works well on MIDI, but SFX cannot be verified since it will crash instantly after game open/start a new save slot( the softstar logo/splash screen/select screen ok).
> 
> <img alt="Image" width="2000" height="1071" src="https://private-user-images.githubusercontent.com/58222/640004687-12291cab-f7df-4209-b671-51d3dcfb8eee.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODc1Mzg5MjksIm5iZiI6MTc4NzUzODYyOSwicGF0aCI6Ii81ODIyMi82NDAwMDQ2ODctMTIyOTFjYWItZjdkZi00MjA5LWI2NzEtNTFkM2RjZmI4ZWVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODI0VDAyMzAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmMWVmMjE0NWJjNGIyMjQ4MWExODM1MjYyM2VmNmY5NWZmMWEwZGJmNzQ3Zjk3OTU4MmVmNTMxNjczNjAwNmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.KbpOeLk5froxfz4qwYsgTEPdt21bDsoWjQqz4jeTwmo">
> On My PC, FreeDOS 1.3( but fdnpkg upgraded to lastest packages ) with CSM: on both DPMI MIDI works well , but SFX broken ( verifed same binary with virtualbox ) . Both tested on HDA and CMI8738, same result. Additional info: if I run `xmsres /l 15` before pal.exe, the result will change, more like vbox situation: SFX will work again with HDPMI ver, and game will crash with VDPMI ver, and same Error 05 will be reported( only the seg addr was changed ).

I have tested on virtualbox and it works, I'm trying to find the problem, any information will help.
e.g. Is there any other drivers/TSRs you installed? what's the DOS version used in virtualbox? what's the option used for VDPMI? what's the result when using /XMS2MEM=15 for VDPMI?

**Comment by palxex** [comment]
> I have tested on virtualbox and it works, I'm trying to find the problem, any information will help. e.g. Is there any other drivers/TSRs you installed? what's the DOS version used in virtualbox? what's the option used for VDPMI?

My VirtualBox FreeDOS setup that I tried yestoday was almost same as my PC's, also a 1.3 setup. Relative lines in fdconfig.sys
```
7?DEVICE=C:\FDOS\DRIVERS\SBEMUV\HIMEMX.EXE
7?DEVICEHIGH=C:\FDOS\DRIVERS\SBEMUV\VDPMI.EXE
```
Other TSRs in `mem /c`:
```
  SYSTEM      86,464   (84K)     10,512   (10K)     75,952   (74K)
  HIMEMX       2,192    (2K)      2,192    (2K)          0    (0K)
  VDPMI        4,512    (4K)      4,512    (4K)          0    (0K)
  COMMAND      3,376    (3K)          0    (0K)      3,376    (3K)
  SBEMU          624    (1K)          0    (0K)        624    (1K)
  SHSUCDX     11,024   (11K)          0    (0K)     11,024   (11K)
  CTMOUSE      3,104    (3K)          0    (0K)      3,104    (3K)
  FDAPM          928    (1K)          0    (0K)        928    (1K)
```

Today I've also tried a MSDOS7 setup with no any tsr loaded, verified same problem(error 05 when load rpg) occured there.  
```
  SYSTEM      17,024   (17K)     13,680   (13K)      3,344    (3K)
  HIMEMX       2,192    (2K)      2,192    (2K)          0    (0K)
  VDPMI        4,512    (4K)      4,512    (4K)          0    (0K)
  COMMAND      7,296    (7K)          0    (0K)      7,296    (7K)
  IFSHLP       2,864    (3K)          0    (0K)      2,864    (3K)
  Free       756,352  (739K)    634,688  (620K)    121,664  (119K)
```

> what's the result when using /XMS2MEM=15 for VDPMI?

On VirtualBox: with this option sometimes( about 50%?) SBEMU will stuck infinitely when loading sound font, both on FreeDOS and MSDOS in VirtualBox. When it works, the game just report same Error 05. 
On real hardware: sound font loads fine every time, and the game will not crash. SFX doesn't play—only a faint noise remains; it's subtle, not glaring like when HDPMI lacks `xmsres /l 15`.

**Comment by crazii** [comment]
Thanks, I will try FreeDOS first. I never tested VDPMI on FreeDOS yet, MSDOS7.1 & 6.22 was used.

**Comment by palxex** [comment]
@crazii  recently I've discovered another quirk in PAL's MPU401.DRV. When switching tracks, the driver does not clear the channel pressure (Aftertouch) state, causing it to bleed into the next track. I've made [a modified version of the driver](https://github.com/palxex/palresearch/discussions/23), but I think its would be great that  SBEMU able to handle it original driver too, since most players will never get to use my patched version anyway.

Due to the driver's internal behavior, I noticed a potential hook: every time it switches tracks, it sends a single byte 0x5 (All Channel Clear) to I/O port 0x331, which was previously ignored by SBEMU (and most other emulators, like DOSBox and VSBHDA). I'm well aware this is non‑standard, but it looks like we could use this as a reliable trigger to clear the pressure state on our side.

Relevant code snippet is attached below, I've tested it works at least under VirtualBox:

```patch
diff --git a/sbemu/vmpu.c b/sbemu/vmpu.c
index 68857e5..336c6dd 100644
--- a/sbemu/vmpu.c
+++ b/sbemu/vmpu.c
@@ -194,6 +194,20 @@ static void VMPU_Write(uint16_t port, uint8_t value)
             midi_message_cntr = 0;
             midi_mpu_status &= ~0x80;
         }
+        if (value == 0x05) {
+            /* MPU-401 "Stop" (all channels clear): re-purpose as CC#121
+             * Reset All Controllers on all 16 channels, so channel pressure
+             * (aftertouch) does not bleed into the next track. */
+            if (!midi_in_sysex) {
+                //need handle incomplete event too? consider later
+                for (uint8_t ch = 0; ch < 16 && midi_ptr + 3 <= 4092; ch++) {
+                    midi_buffer[midi_ptr++] = 0xB0 | ch;
+                    midi_buffer[midi_ptr++] = 0x79;
+                    midi_buffer[midi_ptr++] = 0x00;
+                }
+                midi_available_ptr = midi_ptr;
+            }
+        }
     }
     else
     {
```

I know this is quite hacky – not to mention that sending Intelligent-mode commands while the device is in UART mode is already outside the specification anyway. Still, I thought I'd share it in case you find it useful. Whether to implement it is entirely up to you – I'm happy to go with your decision.

**Comment by crazii** [comment]
Thanks, The truth is, I'm no expert on MPU stuff, almost a noob :)
I think it doesn't hurt to add that, as for the last time fix for PAL when I googled, I found this 05 command for port 331 is not processed, but it is temporarily ignored. I checked TSF as a function called "stop all channel" that could be called.

BTW about the game starting error, I didn't find it in FreeDOS/MSDOS in virualbox, but on a real machine HP T5700 thin client, there is a weird similar error (perhaps error 05) before the startup logo animation, I made some improvements and it's gone.  and now the SFX is OK.

Does your virtualbox running in full acceleration? i.e. there is a "tortoise" logo at the right bottom of your vm window? what the RAM of your virtualbox vm? I need a env to reproduce the problem to test & debug. maybe I can try virtualbox on another PC.

For your test on real PC, it's also weird, I need more tests and think about that. is there any sound fx when using OPL?

**Comment by palxex** [comment]
> Thanks, The truth is, I'm no expert on MPU stuff, almost a noob :) I think it doesn't hurt to add that, as for the last time fix for PAL when I googled, I found this 05 command for port 331 is not processed, but it is temporarily ignored. I checked TSF as a function called "stop all channel" that could be called.

I'm not a MIDI expert either :) I'm just learning from the previous mentioned friend, and AI. My analysis is mostly based on his feedback.

> BTW about the game starting error, I didn't find it in FreeDOS/MSDOS in virualbox, but on a real machine HP T5700 thin client, there is a weird similar error (perhaps error 05) before the startup logo animation, I made some improvements and it's gone. and now the SFX is OK.

Cool! Could you please share the build so I can test it?

> Does your virtualbox running in full acceleration? i.e. there is a "tortoise" logo at the right bottom of your vm window? what the RAM of your virtualbox vm? I need a env to reproduce the problem to test & debug. maybe I can try virtualbox on another PC.

It has the tortoise icon, since I rely on WSL2 heavily, which requires Hyper-V to be turned on. My CPU is an AMD zen3 one, not sure whether it matters. Let me attach the vbox file here.
[vbox98-anonymized.zip](https://github.com/user-attachments/files/31504547/vbox98-anonymized.zip)
I have tried reducing the RAM to 32 MB in virtualbox, but the problem has not changed.

> For your test on real PC, it's also weird, I need more tests and think about that. is there any sound fx when using OPL?

Yes, I just tried it, and PAL.exe still cannot play VOC correctly. On the other hand, in SDLPal's softsynth backends (though they need a modern PC), both music and SFX play well under SBEMU. I think that is because SDLPAL's DMA buffer is allocated in conventional memory, which is enforced by the SoundBlaster backend of SDL3 DOS. This suggests that the root cause still lies in the DMA buffer location. Could you please give me a debug build that logs the physical addresses of all XMS allocations?

**Comment by crazii** [comment]
Here is the debug version of vdpmi, you need specify /LL=2 and /LOG=2 to generate a vdpmi.log in the folder containing vdpmi.exe.
/? has some description on those two options. alternatively, you can use COM output by specify /LOG=1
Note that the log is not in 'append' mode so next restart it'll be probably overwritten.

[vdpmi.zip](https://github.com/user-attachments/files/31515794/vdpmi.zip)

**Comment by palxex** [comment]
[VDPMI.LOG](https://github.com/user-attachments/files/31522949/VDPMI.LOG)
This was logged with `vdpmi /LL /LOG=2`, while with `vdpmi /LL /LOG=2 /XMS2MEM=15` pal.exe will crash( no longer error 05, tested twice ). Wish it will be helpful.

**Comment by palxex** [comment]
A further investigation on MIDI processing: @Cacodemon345 is right that TSF doesn't implement modulator loading, and digging into that gap turned up a new vibrato problem.

VSBHDASF's modified TSF that SBEMU/VSBHDA used is the only one that implements vibrato — upstream TSF doesn't implement modulators, so it had to hardcode a 50-cent amount for the two VibLFO->Pitch (CC#1/mod wheel and channel pressure) defaults to work. That matches the SF2.01 standard, so soundfonts that leave it untouched are fine. But the spec also lets a soundfont override these amounts per-region, and that override is currently lost. Soundbanks that explicitly disable this vibrato (e.g. the sbemusf.sf2 I just uploaded, which sets all the VibLFO->Pitch amount to 0) end up producing an unwanted vibrato anyway.

I've finished an [implementation](https://github.com/sdlpal/sdlpal/commit/1ddead) with AI assistance that only implements the two modulators (CC#1 and channel pressure) the modified TSF needed. It follows fluidsynth's behavior for these two modulators, so it should match what the spec intends, and I've confirmed in SDLPAL that the unwanted vibrato no longer appears with the mentioned soundbank. Just letting you know in case it's useful for your side.

**Comment by crazii** [comment]
> [VDPMI.LOG](https://github.com/user-attachments/files/31522949/VDPMI.LOG) This was logged with `vdpmi /LL /LOG=2`, while with `vdpmi /LL /LOG=2 /XMS2MEM=15` pal.exe will crash( no longer error 05, tested twice ). Wish it will be helpful.

Thanks for the log.
Yes it helps, it seems the very initial xms allocation is in high address ( I don't know why this happens, as I tested the address PING-PONG always started in low address), and then the max address of low xms memory is not clamped to 16M. BTW I haven't check xmsres yet but this functionality of VDPMI may conflict with xmsres, so they better not used together.
Here is an update:

[vdpmi.zip](https://github.com/user-attachments/files/31537333/vdpmi.zip)


> I've finished an [implementation](https://github.com/sdlpal/sdlpal/commit/1ddead) with AI assistance that only implements the two modulators (CC#1 and channel pressure) the modified TSF needed. It follows fluidsynth's behavior for these two modulators, so it should match what the spec intends, and I've confirmed in SDLPAL that the unwanted vibrato no longer appears with the mentioned soundbank. Just letting you know in case it's useful for your side.

I don't understand those code :) , but it is OK as long as it does not crash nor creates new sound artifacts. It's a good improvement, so why not?  Does it improves the original PAL?  I think it is better to ask @Cacodemon345 for help, to review the code.

**Comment by palxex** [comment]
> [vdpmi.zip](https://github.com/user-attachments/files/31537333/vdpmi.zip)

Thanks for new build! But I failed run it on real PC
1. this build will stuck when launch without any parameters, both on MSDOS7 and FreeDOS.
2. when launched with `/LL /LOG=2`( I dont have a second PC to test serial output recently ), it frequently( > 90% ) crash with this error:
```
./core/vm/vm_drv.c(373): in function `VM_DRV_IRQ_Handle`
assertion failed: irq == PIC_INVLD_IRQ
Exiting with error: 2
VDPMI VM86 Exception: 06, Error: 0000
```
Not sure which IRQ it means? I've changed BLASTER's I variable to 2,5,7,9, but none of them fixed it.
3. Due to 2. I just entered game once by luck, but in my memory that time the sfx is still crappy. But I forgot to backup that log, and I failed to repeat in the next about 10 times, so the attached log is just a normal crash one( freedos load vdpmi then sbemu then pal then crash )

[VDPMI.LOG](https://github.com/user-attachments/files/31544198/VDPMI.LOG)

On VirtualBox:
1. stuck when launch without any parameters problem gone .
2. /xms2mem=15 stuck problem gone
3. but the loading game error 05 still there.

[VDPMI.LOG](https://github.com/user-attachments/files/31545967/VDPMI.LOG)

> Does it improves the original PAL? I think it is better to ask [@Cacodemon345](https://github.com/Cacodemon345) for help, to review the code.

Yes, in fact it was meant to fix MIDI vibrato problem in SBEMU/VSBHDA, just implemented in SDLPAL first for testing. I've just tested new SBEMU(both HDPMI and VDPMI version)/VSBHDA built with this patch and found the problem gone too.
Thanks for your advice, I'm going to create a PR in VSBHDASF and request @Cacodemon345 review it.

**Comment by Cacodemon345** [comment]
That patch looks fine at a first glance to be honest.

**Comment by palxex** [comment]
> That patch looks fine at a first glance to be honest.

Thanks for the review:)

**Comment by crazii** [comment]
> ```
> ./core/vm/vm_drv.c(373): in function `VM_DRV_IRQ_Handle`
> assertion failed: irq == PIC_INVLD_IRQ
> Exiting with error: 2
> VDPMI VM86 Exception: 06, Error: 0000
> ```
> 
> Not sure which IRQ it means? I've changed BLASTER's I variable to 2,5,7,9, but none of them fixed it.

It's an assertion only exist in debug build, and triggered on spurious IRQs, it's invalid and will be removed.


The log from your virtualbox seems fine, all the xms allocations are good now. but I don't know the cause of the error yet.

But the log from you real PC, it looks like a mess: although the first 2 or 3 xms allocation give memory below 16M, then after that, until all xms memory is exhausted, there is no available memory below 16M. I believe there's some driver eating xms in MBs. 

Can you comment out all other drivers (FDAPM, SHSUCDX etc. maybe one by one) in fdconfig.sys on your real PC to exclude that possibility? or you have you tried /XMS2MEM=15 on PC? I think 15M reservation is enough and I've never seen drivers uses that much memory except ramdisks like RAMDRIVE/RDISK etc.
Meanwhile I should use a new hack to prevent the memory occupation.

Here is a build with the assertion failure removed: and you probably needs /i7 (IRQ7) for this debug build because irq5 might be masked out.

[vdpmi.zip](https://github.com/user-attachments/files/31580855/vdpmi.zip)

**Comment by palxex** [comment]
> It's an assertion only exist in debug build, and triggered on spurious IRQs, it's invalid and will be removed.

Sure, today's build always able to enter the game.

> The log from your virtualbox seems fine, all the xms allocations are good now. but I don't know the cause of the error yet.
> 
> But the log from you real PC, it looks like a mess: although the first 2 or 3 xms allocation give memory below 16M, then after that, until all xms memory is exhausted, there is no available memory below 16M. I believe there's some driver eating xms in MBs.

> Can you comment out all other drivers (FDAPM, SHSUCDX etc. maybe one by one) in fdconfig.sys on your real PC to exclude that possibility? or you have you tried /XMS2MEM=15 on PC? I think 15M reservation is enough and I've never seen drivers uses that much memory except ramdisks like RAMDRIVE/RDISK etc. Meanwhile I should use a new hack to prevent the memory occupation.
> 
> Here is a build with the assertion failure removed: and you probably needs /i7 (IRQ7) for this debug build because irq5 might be masked out.
> 
> [vdpmi.zip](https://github.com/user-attachments/files/31580855/vdpmi.zip)

I've commented out all other drivers, but currently sfx still not plays under VDPMI version on my PC. 
Now `mem /c` shows
[memc.log](https://github.com/user-attachments/files/31590507/memc.log)
vdpmi still stuck without `/ll /log=2`; log with them

[VDPMI829.LOG](https://github.com/user-attachments/files/31590534/VDPMI829.LOG)

with `/xms2mem=15 /ll /log=2`:

[VDPMI.LOG](https://github.com/user-attachments/files/31590546/VDPMI.LOG)

An idea: the SBEMU/HDPMI version handles SFX fine with `xmsres /l 15`, even with all those drivers loaded. That probably means the problem is with the XMM.

My setup for HDPMI: JEMMEX, QPIEMU, HDPMI32I (from sbemu), xmsres (from vsbhda), then sbemu.
For VDPMI: HIMEMX, VDPMI (both from the sbemu vdpmi package), then sbemu.

So maybe HIMEMX is the issue? But I can't test that with JEMMEX because VDPMI won't start when an EMM is there (even with /EMS=0). Any switch or hack to make VDPMI ignore the EMM? Not really familiar with low-level stuff such as DPMI, so sorry if I'm way off. :)

**Comment by crazii** [comment]
> So maybe HIMEMX is the issue? But I can't test that with JEMMEX because VDPMI won't start when an EMM is there (even with /EMS=0). Any switch or hack to make VDPMI ignore the EMM? Not really familiar with low-level stuff such as DPMI, so sorry if I'm way off. :)

VDPMI needs its internal v86 monitor to work properly, and that conflicts with any other EMMs. So there is no way to do that, it's just the way how it is designed and implemented.

Are you using the latest HIMEMX from github? or the one shipped with FD1.3? I tested FD1.3 with HIMEMX installed from FD1.3's liveCD, it works fine. FD1.4 also tested fine. The latest HIMEMX from github also works fine as I last tried in MSDOS7.  You may try to update HIMEMX to the latest release on github: https://github.com/Baron-von-Riedesel/HimemX/releases/tag/v3.40
and if you're using MSDOS (or you have that option), you can try HIMEM.SYS.

The new log is same as the last, and currently I out of ideas, maybe further approaches could be taken after you test with the latest HIMEMX. Or it can be fixed in the long term.

**Comment by palxex** [comment]
> VDPMI needs its internal v86 monitor to work properly, and that conflicts with any other EMMs. So there is no way to do that, it's just the way how it is designed and implemented.

Just a sudden idea, never mind:)

> Are you using the latest HIMEMX from github? or the one shipped with FD1.3? I tested FD1.3 with HIMEMX installed from FD1.3's liveCD, it works fine. FD1.4 also tested fine. The latest HIMEMX from github also works fine as I last tried in MSDOS7. You may try to update HIMEMX to the latest release on github: https://github.com/Baron-von-Riedesel/HimemX/releases/tag/v3.40 and if you're using MSDOS (or you have that option), you can try HIMEM.SYS.

Previous I was always testing with HIMEMX.EXE in SBEMU_VDPMI.zip. 

> The new log is same as the last, and currently I out of ideas, maybe further approaches could be taken after you test with the latest HIMEMX. Or it can be fixed in the long term.

OK... I've just tested both combinations (FreeDOS + HIMEMX 3.4, and MS-DOS + HIMEM), and no luck:( Still, FM/MIDI music works, but SFX doesn't. So it seems like some edge case specific to this PC build. Let's wait for further inspiration.

For MSDOS test, I just used IO.SYS, HIMEM.SYS, etc. from  [Wengier's MSDOS710 bootdisk](https://www.cn-dos.net/msdos71/msdos71b.zip), so that anyone can reproduce my setup.

**Comment by crazii** [comment]
> Previous I was always testing with HIMEMX.EXE in SBEMU_VDPMI.zip.

That one was the latest, I get it from github.

I will do more test to check if anything has been missed.
Or I can try to bypass the XMM and manage the lower XMS totally. Actually there was a long term plan to implement an built-in XMM, but with the current stability, I don't think it's the time.

**Comment by crazii** [comment]
I think I know why this happens:
You pc has multiple memory holes in high memory that makes multiple fragmentations.

For xms reservation for xms memory with no fragmentations, there's a "hole" occupied by VDPMI and makes 2 fragmentations, the xmm will literate each fragmentation during multiple allocations, causing a ping-pong style between the 2 fragmentations.

When your PC have more pci devices that needs memory address for mmio, there could be many fragmentations in the high memory.  So it's not a ping-pong style but using each fragmentation in turn, during multiple allocations.

That exhausts the xms handles not the memory.

I will figure out a way for it. For fast tests, there's an option for himemx.exe/himem.sys to set handle limit, you can try it with maximum; it might not work yet because VDPMI has a limit for retry count, but at least it will definitely generate a better log that more allocations will succeed in low mem below 16M.

**Comment by crazii** [comment]
I can't think of any good way to solve it except to bypass the XMM and manage the lower XMS totally. 

I can image how xmsres handles the problem, probably bypass the XMM too but in a very simple way.
I may need to check if it has a better way than I thought.

EDIT: checked source of xmsres and I don't think it's suitable for VDPMI. it exhausts all high memory above 16M, by merging allocations to save xms handles. and leave the left lower 16M for programs.

**Comment by palxex** [comment]
> I think I know why this happens: You pc has multiple memory holes in high memory that makes multiple fragmentations.
> 
> For xms reservation for xms memory with no fragmentations, there's a "hole" occupied by VDPMI and makes 2 fragmentations, the xmm will literate each fragmentation during multiple allocations, causing a ping-pong style between the 2 fragmentations.
> 
> When your PC have more pci devices that needs memory address for mmio, there could be many fragmentations in the high memory. So it's not a ping-pong style but using each fragmentation in turn, during multiple allocations.

> 
> That exhausts the xms handles not the memory.
> 
> I will figure out a way for it. For fast tests, there's an option for himemx.exe/himem.sys to set handle limit, you can try it with maximum; it might not work yet because VDPMI has a limit for retry count, but at least it will definitely generate a better log that more allocations will succeed in low mem below 16M.

Excellent analysis! I've tried add ` /NUMHANDLES=128` to parameters of HIMEMX.EXE, now both in FreeDOS and MSDOS, PAL.EXE plays SFX well.

Log for FreeDOS:
[VDPMI.LOG](https://github.com/user-attachments/files/31606346/VDPMI.LOG)
MSDOS:
[VDPMI.LOG](https://github.com/user-attachments/files/31606366/VDPMI.LOG)

Meanwhile, MSDOS with DEVICE=HIMEM.SYS /NUMHANDLES=128 results in error 05 in PAL.EXE. It doesn't matter now, but maybe this implies that this PAL error is related to XMS handling, so just leave it here for future researchers.

> EDIT: checked source of xmsres and I don't think it's suitable for VDPMI. it exhausts all high memory above 16M, by merging allocations to save xms handles. and leave the left lower 16M for programs.

Agree — it's okay as a hacky bypass, but not a good approach for future design.

**Comment by crazii** [comment]
Glad to hear that. in short is an ABABAB style changed to something like ABCDABCDABCD style.

In FreeDOS, there're still allocations above 16M, "pa=bdc9b000 pe=bdca8000" "pa=a82a000 pe=a83d800" "pa=a98c000 pe=a9b6000" "pa=bdc9b000 pe=bdca8000" "pa=aaf5b000 pe=aaf63800",

In MSDOS, "pa=abca000 pe=ac1d000" "pa=ab117000 pe=ab141000" "pa=bdc9b000 pe=bdca8000" "pa=aaf5b000 pe=aaf63800"

Those are exceeding the internal retry limit count in VDPMI (64).

Error 05 might be caused by those high addresses. I'm not sure because in virutalbox all allocations seems OK but there's still an error 05. LOL

I'm planning to take over the low 16M xms entirely. There is no other better way to do it. Depending on the max handle parameter is not a good solution.

Use the method of xmsres "on the fly": take all of high address memories, before each an every allocation request (and release some on request failure). that could also work if all information are pre-gathered, and high address allocations are pre-merged for only once.

I might need your help to test this case later after I finishes this, Thanks! :D

**Comment by palxex** [comment]
Glad to be helpful:) Feel free to ping me once done.

**Comment by crazii** [comment]
Hi @palxex ,

I've added an built-in XMM (extended memory manager) that support the allocation strategies, which defaults to bottom-up: always allocate from the lowest address available.
I've tested PAL in MSDOS7 and FreeDOS in virtual box and it works.

the builtin XMM is optional, so there are two test cases:
1) installing HIMEMX.EXE/HIMEM.SYS before VDPMI. there will be a warning on installing vpdmi and it can be ignored.
2) install VDPMI without HIMEMX.EXE/HIMEM.SYS

both cases works for me. I'll do more tests on real machines. you can do a quick test to see if it works.

updated vdpmi with infinite error msg:
2nd update with logfile fix:

[vdpmi-xmm.zip](https://github.com/user-attachments/files/31883878/vdpmi-xmm.zip)

**Comment by palxex** [comment]
## Test Results: VDPMI-XMM + SBEMU on My PC and VirtualBox

### Summary
| Environment |with HIMEMX | Result |
| :--- | :--- | :--- |
| MSDOS7 (my PC) | No | SBEMU floods screen, then reboots — no log file |
| FreeDOS (my PC) | No | Same as above |
| MSDOS7  (my PC) | Yes (`/NUMHANDLES=128`) | PAL.EXE crashes with Error 05 after loading a saved game |
| FreeDOS (my PC) | Yes (`/NUMHANDLES=128`) | Same as above |
| FreeDOS (VirtualBox) | No | Screen flooding (identical to My PC) — too fast to capture |
| FreeDOS (VirtualBox) | Yes | PAL.EXE crashes with Error 05 after loading a saved game |

### Detailed Results

**On my PC (both MSDOS7  and FreeDOS)**  
- Without HIMEMX: Loading VDPMI 0.3 alone causes the console to flood with continuous output, scrolling too fast to determine whether the output originates from VDPMI or SBEMU. This is followed by a system reboot. No `VDPMI.LOG` is generated. All logs attached below were produced with `HIMEMX /NUMHANDLES=128` followed by `VDPMI /LL /LOG=2` (VDPMI without parameters still hangs).
- With HIMEMX:
  - when able to enter game: With the above configuration, the game crashes with Error 05 immediately after loading a saved game — consistent with behavior previously observed in VirtualBox. The status of the SFX playback fix could not be confirmed due to the crash.
  - when crash before DOS Shell usable: sometimes it happens( less than 50% in my feeling ); but the crash manifests in multiple distinct patterns, see [scrnshots.zip](https://github.com/user-attachments/files/31837512/scrnshots.zip) for examples.

**Logs (My PC)**:  
FreeDOS: [VDPMI.LOG](https://github.com/user-attachments/files/31833910/VDPMI.LOG)  
MSDOS7: [VDPMI.LOG](https://github.com/user-attachments/files/31833897/VDPMI.LOG)

**On VirtualBox (FreeDOS)**  
- Without HIMEMX: Screen flooding appears identical to real PC, though the latter one was too fast to confirm with certainty.  
  ![Image](https://github.com/user-attachments/assets/a3c6d4ac-6a31-458d-9ccb-d302e3b1be70)  
- With HIMEMX: Error 05 occurs at the same location as previous tests.

**Logs (VirtualBox)**:  
[VDPMI.LOG](https://github.com/user-attachments/files/31833953/VDPMI.LOG)

**Comment by palxex** [comment]
Further investigation into Error 05 of PAL.EXE: this error code maps to QBasic runtime error 5 (`B$ERR_FC`), meaning "Illegal function call". However, static analysis of PAL.EXE and tracing in DOSBox both suggest that the execution flow should not normally reach the segment:offset where the error is triggered when loading a saved game. This implies prior memory corruption or some other form of data/state corruption, which I have not been able to thoroughly investigate.

**Comment by crazii** [comment]
Thanks.

It seems the xmm in system level does not work properly.
same xmm code also used with external xmm for the low xms taken over, but 05 error not fixed. Error 05 might be related to xms memory copy or timing, I will check both situations.

Still a long way to go.

Btw /LL=2 could generate more info for logs, but I don't think it's needed for now, maybe on future fixes.

Also /NUMHANDLES=128 is not needed for this build, but it does matter.

**Comment by crazii** [comment]
I've updated the zip (https://github.com/crazii/SBEMU/issues/148#issuecomment-5536547202) that fixes the infinite error msg.

I think for this time, to make things easy, you can test it only in virtual box, and only without himemx/himem.

BTW last time I saw your vbox file (https://github.com/crazii/SBEMU/issues/148#issuecomment-5436437726), that you use file for COM (VM Settings/Serial Ports). I think you can use /LOG=1 for vdpmi to output the log to COM and let vbox generate a file and upload it, which may be more helpful than a fast scrolling screen.

**Comment by palxex** [comment]
Thanks for the advice 🙂 — the problem is that I’ve never managed to get VirtualBox’s raw file dump for RS232 to work (the file never actually gets generated). The logs I’ve been providing above were all generated using `VDPMI /ll /log=2` instead.
VirtualBox 7.2.12 on Windows 11 x64 with WHP test results:
- with himemx( no more /numhandler=128 ):
  [VDPMI.LOG](https://github.com/user-attachments/files/31868214/VDPMI.LOG)
  PAL still error 05
- without himemx:
  [VDPMI.LOG](https://github.com/user-attachments/files/31868299/VDPMI.LOG)
  DOS shell will stuck after sbemu loading, so cannot continue testing.
<img width="2342" height="1812" alt="Image" src="https://github.com/user-attachments/assets/06b26497-3ad1-4e29-bd07-13ef7f43e150" />

**Comment by crazii** [comment]
Thank you!

The log is too simple, and the exception 06 is "invalid opcode" which usually means the execution has gone wild at random memory locations that cannot be interpreted as CPU instructions.

Can you do a simple test without installing himemx, and set /LL=2 for vdpmi?

the COM to raw file works for me, on windows10 vbox 7.2.6, and on Linux. I don't know what happened for your system. I have a win11 but don't remember whether using that feature on it.

EDIT: the logfile without himemx, its content indicates that himemx is installed.

**Comment by palxex** [comment]
> EDIT: the logfile without himemx, its content indicates that himemx is installed.

Sorry about that:( Seems when crashing, the logging file was not even touched, so its content was actually the last time logged, which was logging with himemx.

> the COM to raw file works for me, on windows10 vbox 7.2.6, and on Linux. I don't know what happened for your system. I have a win11 but don't remember whether using that feature on it.

Found the problem, seems another service I installed blocks writing. After stops it now the serial log could be generated, and its now not crashing( but revert to file logging will still crash with exception 06 ). Now only error 05 in PAL leaves.
Serial log with `vdpmi /LL=2 /LOG=1` without HIMEMX  on VirtualBox:
[rs232.txt](https://github.com/user-attachments/files/31873008/rs232.txt)

**Comment by crazii** [comment]
Thanks! I thought that the log file is not even created: the last screen message "Logfile TB XXX" is right during logfile init, I think it crashes after that and before creating the file.

I will focus on the logfile problem and the PAL error 05 problem.

**Comment by crazii** [comment]
About error 05 in virtualbox, I'm unable to find any clues, but I've built a exe that will traces "all interrupts" and dump registers if some interrupts raises CFLAG. It may help to find the cause, e.g. the last interrupt number used before the error 05 report should be checked. The log is heavy and could be a little bit slow, but since that you can use COM raw file, I think it is OK to use that built, it's updated in the last upload: https://github.com/crazii/SBEMU/issues/148#issuecomment-5536547202

The logfile problem indeed is a bug when the XMM enabled in system level, it is exposed by the code of the logfile.
the problem is fixed so maybe you can test it on your PC with /LL=0 (disable all logs)

EDIT: 2 critical bugs that might affect installation are fixed, maybe it works for your friend now.

**Comment by palxex** [comment]
- VirtualBox with serial logging: `vdpmi /LL=2 /LOG=1` produced [rs232.txt](https://github.com/user-attachments/files/31913415/rs232.txt) but the game became extremely slow after loading, making it impossible to verify whether SFX works.
- VirtualBox with `vdpmi /LL=0 /LOG=1`: still crashes with Error 05.
- My Real PC with `vdpmi /LL=0 /LOG=2`: no log file was generated, and Error 05 still occurs.
- My friend's laptop( i5+H81 ) could enter VDPMI/SBEMU and play PAL without error 05 but also no SFX, but he said SDLPal was not stable, softsynth audio may stuck sometimes.

**Comment by crazii** [comment]
> * VirtualBox with serial logging: `vdpmi /LL=2 /LOG=1` produced [rs232.txt](https://github.com/user-attachments/files/31913415/rs232.txt) but the game became extremely slow after loading, making it impossible to verify whether SFX works.
> * VirtualBox with `vdpmi /LL=0 /LOG=1`: still crashes with Error 05.
> * My Real PC with `vdpmi /LL=0 /LOG=2`: no log file was generated, and Error 05 still occurs.
> * My friend's laptop( i5+H81 ) could enter VDPMI/SBEMU and play PAL without error 05 but also no SFX, but he said SDLPal was not stable, softsynth audio may stuck sometimes.

The last call made before error message is calling INT 21h for UMB links. you may remove umb by set `DOS=HIGH` only in fdconfig.sys to see if works.
In the mean time, I will have to write full test cases on UMB/EMS/XMS to find the problem.

**Comment by palxex** [comment]
> The last call made before error message is calling INT 21h for UMB links. you may remove umb by set `DOS=HIGH` only in fdconfig.sys to see if works. In the mean time, I will have to write full test cases on UMB/EMS/XMS to find the problem.

With DOS=HIGH only, `DOS=UMB/DOSDATA=UMB` was commented,  still error 05.
New serial log: [rs232.txt](https://github.com/user-attachments/files/31979118/rs232.txt)

**Comment by crazii** [comment]
> > The last call made before error message is calling INT 21h for UMB links. you may remove umb by set `DOS=HIGH` only in fdconfig.sys to see if works. In the mean time, I will have to write full test cases on UMB/EMS/XMS to find the problem.
> 
> With DOS=HIGH only, `DOS=UMB/DOSDATA=UMB` was commented, still error 05. New serial log: [rs232.txt](https://github.com/user-attachments/files/31979118/rs232.txt)

Thank you, then please wait for me to test more on this.
